Interpolation events are now per target.

markscheel opened the pull request:
## Proposed changes

Previously there was only one Interpolation event, so (e.g.) if you wanted to have a different
event for finding each of three apparent horizons AhA, AhB, and AhC, there was no way to do it.

Now the interpolation event is templated by the InterpolationTarget, and you specify one event
per target in the input file. Furthermore, the name of the event in the input file is the name of the
InterpolationTarget.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

markscheel committed:
Events::Interpolate templated on TargetTags.

So now instead of a single interpolation event, you can have
multiple interpolation events, one for each tag.  And furthermore,
when you specify the Event in the input file, you specify it by the
name of the InterpolationTargetTag.

markscheel committed:
Factory.hpp: use option_name, not short_name.

Now one can use a 'name' function to override
the name used in the yaml file.

moxcodes commented on tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml:
```diff
@@ -54,7 +54,7 @@ EventsAndTriggers:
       N: 5
       Offset: 2
-  : - Interpolate
+  : - Horizon
```
Is it possible to set the option name string for the event so it is more obvious this is an interpolation?

fmahebert replied:
In particular, it would be nice to disambiguate between interpolating to some horizon-like strahlkorper, vs. performing an iterative horizon find.

markscheel replied:
Yes, you can set the name to whatever you want.  It uses the name of the InterpolationTargetTag (or a name function inside InterpolationTargetTag), which you can set to whatever you want. Maybe I shouldn't have picked the name 'Horizon'.  Maybe 'InterpolateToKerrHorizon'?

fmahebert replied:
Personally I find this suggestion much clearer 👍

nilsdeppe replied:
was this resolved? The input file still says `Horizon`

fmahebert commented on src/NumericalAlgorithms/Interpolation/Interpolate.hpp:
```diff
@@ -36,6 +37,8 @@ template <size_t VolumeDim> class ElementIndex;
 template <typename Metavariables>
 struct Interpolator;
+template <typename Metavariables, typename Tag>
+struct InterpolationTarget;
```
[nitpick] alphabetize

fmahebert commented on src/NumericalAlgorithms/Interpolation/Interpolate.hpp:
```diff
@@ -44,26 +47,30 @@ struct InterpolatorReceiveVolumeData;
-          typename EventRegistrars =
-              tmpl::list<Registrars::Interpolate<VolumeDim, Tensors>>>
+/// Does an interpolation onto InterpolationTargetTag by calling Actions on
+/// Interpolator and InterpolationTargetTag.
```
[nitpick] I think this would be slightly clearer if rephrased to "... on the Interpolator and InterpolationTargetTag components".

fmahebert reviewed (commented):
Overall LGTM, but here are a few non-binding minor clarity improvements

markscheel committed:
EvolveGeneralizedHarmonic: Rename Horizon->AhA

This is supposed to disambiguate whether we are interpolating to
an apparent horizon or whether we are interpolating to a set of
points that is chosen according to an anaytical horizon in some
coordinate system.  Once we have binaries, there will also be
an AhB and AhC.

markscheel committed:
FishboneMoncriefDisk: Rename Horizon->KerrHorizon.

This is supposed to disambiguate whether we are interpolating to
an apparent horizon or whether we are interpolating to a set of
points that is chosen according to an anaytical horizon in some
coordinate system.

markscheel commented:
Added fixup commits.  The 'Horizon' InterpolationTargetTag has been renamed to 'AhA' in EvolveGeneralizedHarmonic (where an iterative horizon find is being done) and 'InterpolateToKerrHorizon' in EvolveFishboneMoncriefDisk, where a fixed set of points are used.

moxcodes reviewed (commented):
I'm happy for you to squash

fmahebert commented:
I'm also happy for a squash

markscheel commented:
Squashed.  Left the two renaming commits unsquashed since they are separate.  I needed to change InterpolateToKerrHorizon to InterpToKerrHorizon because the former exceeded the 21 character limit.

kidder commented on tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml:
```diff
@@ -54,7 +54,7 @@ EventsAndTriggers:
       N: 5
       Offset: 2
-  : - Horizon
+  : - AhA
```
Does this find the horizon or observe something on it (or both)?  If the former I would prefer FindAhA, if eventually both than I am fine with AhA

markscheel replied:
It does both.  At the end of the horizon find, it calls whatever post_horizon_find_callback has been specified in the InterpolationTargetTag in the metavariables.  The current post_horizon_find_callback does an observation of the horizon area.

kidder commented on tests/InputFiles/GrMhd/ValenciaDivClean/FishboneMoncriefDisk.yaml:
```diff
@@ -87,7 +87,7 @@ Observers:
 
 InterpolationTargets:
-  Horizon:
+  InterpToKerrHorizon:
```
Sorry for the questions, but I've gotten lost...  I just need reminding of what this actually enables, and what capabilities are coming next...  Does this do anything besides interpolate to the target (i.e. is something written to file?)  Is it being done always or never (I don't see it in the triggers)?    If it does something besides interpolate, how do I control what is observed.   Also I think `KerrHorizon` is sufficient since the `InterpTo` is redundant with `InterpolationTargets` .  Also you should use yaml references to keep the mass and spin of the solution and interpolation target the same...

markscheel replied:
It interpolates and calls obsevers (currently writes the mass flux, but eventually will write other things too.)  You control what is being observed via the InterpolationTargetTag in the metavariables, which specifies the observers. Currently there is no trigger in this input file, so interpolation/observation is done never; a trigger needs to be added to do the interpolation/observation.

markscheel replied:
I can change the name and look at how yaml references work.

kidder replied:
Okay, thanks for clarifying!

markscheel committed:
FishboneMoncriefDisk.yaml: use yaml references.

So we don't repeat nubers in the input file.

markscheel commented:
Renamed InterpToKerrHorizon->KerrHorizon (and squashed that commit), use yaml references for mass and spin (new commit).

kidder reviewed (commented):
you can squash

markscheel commented:
Would it be better to keep commits as-is (i.e. no squash), because the last yaml commit does something different than the others?

kidder reviewed (dismissed):
yes, sorry I just assumed a squash was in order...

markscheel commented:
Rebased on develop after all the test refactoring.

nilsdeppe commented on src/Options/Factory.hpp:
```diff
@@ -35,7 +34,7 @@ std::unique_ptr<BaseClass> create_derived(const std::string& id,
   using derived = tmpl::front<CreateList>;
 
-  if (pretty_type::short_name<derived>() != id) {
+  if (option_name<derived>() != id) {
```
Can you add a test for this part of the factory?

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -163,7 +163,7 @@ struct EvolutionMetavars {
   };
 
-  struct Horizon {
+  struct AhA {
```
Not for this PR, but I generally never liked `AhA` and `AhB` because I never know which is the larger hole. Could we switch to something like `LargerAh` or `AhBigger` (and an opposite name) in the future? And then have `AhC` be `AhCommon` or `CommonAh`? I think that would help make the code a lot more accessible.

nilsvu replied:
Yes please :D Perhaps `Primary` and `Secondary`?

markscheel replied:
Ok, I can change in whatever PR actually has two holes.  Note that for equal mass systems, either hole can be slightly larger than the other due to truncation error, so the names will not always be correct.

nilsdeppe replied:
Thanks! Yes, though personally rather have the name be incorrect by truncation error than be unsure about every simulation :)

nilsdeppe reviewed (commented):
I'm happy for you to squash any changes.

markscheel commented:
Added test for Factory and squashed it into Factory commit.

nilsdeppe reviewed (approved):
![image](https://user-images.githubusercontent.com/13531030/77566980-db372800-6e9c-11ea-989b-263eff06b588.png)

Thanks for adding the test!
